File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    nsum = 0
    def bstToGst(self, root: TreeNode) -> TreeNode:
        self.nodeSum(root)
        return root
    
    def nodeSum(self, node: TreeNode):
        if node is None:
            return
        
        global nsum
        
        self.nodeSum(node.right)
        self.nsum += node.val
        node.val = self.nsum
        self.nodeSum(node.left)
        

File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int nsum = 0;
    TreeNode* bstToGst(TreeNode* root) {
        nodeSum(root);
        
        return root;
    }
    
    void nodeSum(TreeNode* node) {
        if(node == NULL) return;
        
        nodeSum(node->right);
        
        nsum += node->val;
        node->val = nsum;
        
        nodeSum(node->left);
    }
};